### PR TITLE
Added RTDETR model to inference

### PR DESCRIPTION
**Pull request by Bhavay-2001**
# Description

PR for #546 

## Type of change

Please delete options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [x] This change requires a documentation update

## Docs

-   [ ] Docs updated? What were the changes:

### CODE REVIEW

**Comment by Bhavay-2001** [comment]
Hi @grzegorz-roboflow, I haven't tested it myself. I got the code from [here](https://huggingface.co/docs/transformers/en/model_doc/rt_detr). I have created the [notebook](https://colab.research.google.com/drive/1dx0FlqB97w7Uo9VCYgdiqAJakKTR2H04?usp=sharing) that test this code. 

I have just tried to convert that code to a file.

**Comment by Bhavay-2001** [comment]
Hi @grzegorz-roboflow, pls let me know what changes needs to be done.

File: inference/models/rtdetr/rtdetr.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+from PIL import Image
+from transformers import RTDetrForObjectDetection, RTDetrImageProcessor
+
+from inference.core.models.base import PreprocessReturnMetadata
+from inference.core.models.roboflow import RoboflowCoreModel
+from inference.core.utils.image_utils import load_image_rgb
+
+DEVICE = "cuda:0" if torch.cuda.is_available() else "cpu"
+
+class RTDETR(RoboflowCoreModel):
```
[the remark this answers is not in the document]
Comment (Bhavay-2001) [reply inference/models/rtdetr/rtdetr.py]: Done

**Comment by Bhavay-2001** [comment]
Hi @probicheaux, can you pls provide a detailed review?
Thanks

**Comment by Bhavay-2001** [comment]
Hi @grzegorz-roboflow, PR is ready for review. Can you please check this. 
Thanks

File: inference/models/rtdetr/rtdetr.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+from typing import Any, Tuple
+
+from PIL import Image
+from transformers import RTDetrForObjectDetection, RTDetrImageProcessor
+
+from inference.core.models.base import PreprocessReturnMetadata
+from inference.models.transformers.transformers import TransformerModel
+from inference.core.utils.image_utils import load_image_rgb
+
+DEVICE = "cuda:0" if torch.cuda.is_available() else "cpu"
```
Comment (probicheaux) [inline inference/models/rtdetr/rtdetr.py:15]: Not needed, self.device should be set on the TransformerModel

File: inference/models/rtdetr/rtdetr.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+class RTDETR(TransformerModel):
+    def __init__(self, *args, model_id=f"", **kwargs):
+        super().__init__(*args, model_id=model_id, **kwargs)
+        self.model_id = model_id
+        self.endpoint = model_id
+        self.api_key = API_KEY
+        self.dataset_id, self.version_id = model_id.split("/")
+        self.cache_dir = os.path.join(MODEL_CACHE_DIR, self.endpoint + "/") # "PekingU/rtdetr_r50vd"
+        dtype = torch.bfloat16
```
Comment (probicheaux) [inline inference/models/rtdetr/rtdetr.py:25]: bfloat16 shouldn't be hardcoded here, bfloat16 is only supported on gpus with compute capability >= 8.0

File: inference/models/rtdetr/rtdetr.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+from inference.core.utils.image_utils import load_image_rgb
+
+DEVICE = "cuda:0" if torch.cuda.is_available() else "cpu"
+
+class RTDETR(TransformerModel):
+    def __init__(self, *args, model_id=f"", **kwargs):
+        super().__init__(*args, model_id=model_id, **kwargs)
+        self.model_id = model_id
+        self.endpoint = model_id
+        self.api_key = API_KEY
```
Comment (probicheaux) [inline inference/models/rtdetr/rtdetr.py:22]: I don't think these 3 lines are needed, this is set on RoboflowInferenceModel

File: inference/models/rtdetr/rtdetr.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        self.dataset_id, self.version_id = model_id.split("/")
+        self.cache_dir = os.path.join(MODEL_CACHE_DIR, self.endpoint + "/") # "PekingU/rtdetr_r50vd"
+        dtype = torch.bfloat16
+        self.model = RTDetrForObjectDetection.from_pretrained(
+            self.cache_dir,
+            torch_dtype=dtype,
+            device_map=DEVICE,
+            revision="bfloat16",
```
Comment (probicheaux) [inline inference/models/rtdetr/rtdetr.py:30]: We can upload float16 weights to a Roboflow project and load from there

File: inference/models/rtdetr/rtdetr.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+class RTDETR(TransformerModel):
+    def __init__(self, *args, model_id=f"", **kwargs):
+        super().__init__(*args, model_id=model_id, **kwargs)
+        self.model_id = model_id
+        self.endpoint = model_id
+        self.api_key = API_KEY
+        self.dataset_id, self.version_id = model_id.split("/")
+        self.cache_dir = os.path.join(MODEL_CACHE_DIR, self.endpoint + "/") # "PekingU/rtdetr_r50vd"
+        dtype = torch.bfloat16
+        self.model = RTDetrForObjectDetection.from_pretrained(
```
Comment (probicheaux) [inline inference/models/rtdetr/rtdetr.py:26]: RTDetrForObjectDetection should be a class property, see https://github.com/Bhavay-2001/roboflow-inference/blob/d3c88f74fdcaac5c29822a7444698b11b78067c8/inference/models/paligemma/paligemma.py#L10
For an example

File: inference/models/rtdetr/rtdetr.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            self.cache_dir,
+            torch_dtype=dtype,
+            device_map=DEVICE,
+            revision="bfloat16",
+        ).eval()
+
+        self.processor = RTDetrImageProcessor.from_pretrained(
```
Comment (probicheaux) [inline inference/models/rtdetr/rtdetr.py:33]: Same comment for class property

**Review by probicheaux — COMMENTED** [review]
Thanks for your contribution, we will upload some weights and test this out. Some unit tests/integration tests will also need to be added

**Comment by yeldarby** [comment]
Hey @Bhavay-2001!  We greatly appreciate your contribution and thank you so much for submitting this PR.

I talked with the team and we are planning on incorporating fine-tuning of RT-DETR more tightly into Roboflow shortly so we need to adapt the inference integration to be compatible with the output of our training process.

This PR is a great start & we’ll likely continue working from it towards a release but we won’t be able to do that for a little while until the backend is more concrete. Best path forward in the meantime would be to use it via a plugin or fork. (Apologies for the inconvenience; we’ll update you when we have more to share on our end!)

**Pull request closed without merging**